### PR TITLE
Add instrumentation for Mysql.Data.

### LINKED ISSUE #113: Add instrumentation for Mysql.Data
# Feature Request

**Is your feature request related to a problem?**

I want to trace mysql executions.

**Describe the solution you'd like:**

Mysql executions should be traced correctly, following [opentelemetry-specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/database.md#mysql).

**Describe alternatives you've considered.**

I've tried [opentelemetry-dotnet-instrumentation](https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation) but for now it's not ready for production.

### PR DESCRIPTION
**Pull request by moonheart**
Fixes #113

Add instrumentation for [Mysql.Data](https://www.nuget.org/packages/MySql.Data/).

### CODE REVIEW

File: src/OpenTelemetry.Contrib.Instrumentation.MySqlData/OpenTelemetry.Contrib.Instrumentation.MySqlData.csproj [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    <PropertyGroup>
+        <TargetFramework>netstandard2.0</TargetFramework>
+        <Description>OpenTelemetry instrumentation for Mysql.Data</Description>
+        <PackageTags>$(PackageTags);distributed-tracing;Mysql.Data</PackageTags>
+        <MinVerTagPrefix>Instrumentation.MysqlData-</MinVerTagPrefix>
+        <IncludeSharedInstrumentationSource>true</IncludeSharedInstrumentationSource>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <PackageReference Include="MySql.Data" Version="8.0.25" />
```
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.MySqlData/OpenTelemetry.Contrib.Instrumentation.MySqlData.csproj:12]: This instrumentation only works for this version and above? Is that the intention?

File: src/OpenTelemetry.Contrib.Instrumentation.MySqlData/OpenTelemetry.Contrib.Instrumentation.MySqlData.csproj [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+      <PackageReference Include="OpenTelemetry" Version="1.1.0-beta3" />
```
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.MySqlData/OpenTelemetry.Contrib.Instrumentation.MySqlData.csproj:13]: Only OpenTelemetry.API dependency is required, not required to add OpenTelemetry

File: src/OpenTelemetry.Contrib.Instrumentation.MySqlData/README.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            .AddMySqlDataInstrumentation()
+            .AddConsoleExporter()
+            .Build();
+    }
+}
+```
+
+For an ASP.NET Core application, adding instrumentation is typically done in
+the `ConfigureServices` of your `Startup` class. Refer to documentation for
+[OpenTelemetry.Instrumentation.AspNetCore](../OpenTelemetry.Instrumentation.AspNetCore/README.md).
```
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.MySqlData/README.md:47]: these links need to be fixed to point to the main repo.

**Review by cijothomas — CHANGES_REQUESTED** [review]
Thanks for your contribution!
Not a MySql expert, but PR looks good from instrumentation standpoint.
Left couple of comments, most about the package dependencies.

**Comment by cijothomas** [comment]
> [CLA Not Signed](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/261573555/115/#/?version=1)
> 
> 
> * [❌](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/261573555/115/#/?version=1) - moonheart The commit ([8c3b983](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/commit/8c3b983d90177514bfcd92c1e4ebaa8a8d189e6b) ,d11896942b9d18f0c462107816ea301a94aa39ff) is not authorized under a signed CLA. [Please click here to be authorized](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/261573555/115/#/?version=1). For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

You need to sign the CLA before any PRs can be accepted.

File: src/OpenTelemetry.Contrib.Instrumentation.MySqlData/MySqlDataInstrumentation.cs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+            MySqlTrace.Listeners.Clear();
+            MySqlTrace.Listeners.Add(this);
+            MySqlTrace.Switch.Level = SourceLevels.Information;
+            MySqlTrace.QueryAnalysisEnabled = true;
+        }
+
+        /// <inheritdoc />
+        public override void TraceEvent(TraceEventCache eventCache, string source, TraceEventType eventType, int id, string format, params object[] args)
+        {
+            switch ((MySqlTraceEventType)id)
```
Comment (cijothomas) [inline src/OpenTelemetry.Contrib.Instrumentation.MySqlData/MySqlDataInstrumentation.cs:49]: Do you need to have this whole body inside a try..catch?

**Comment by moonheart** [comment]
Hello @cijothomas , thanks for your review, I've fixed these problems, please take a look.

**Comment by moonheart** [comment]
Hello @cijothomas , is there any suggestions for this pull request?

File: src/OpenTelemetry.Contrib.Instrumentation.MySqlData/MySqlDataInstrumentation.cs [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+                        this.AfterExecuteCommand();
+                        break;
+                    case MySqlTraceEventType.StatementPrepared:
+                        break;
+                    case MySqlTraceEventType.StatementExecuted:
+                        break;
+                    case MySqlTraceEventType.StatementClosed:
+                        break;
+                    case MySqlTraceEventType.NonQuery:
+                        break;
```
Comment (alanwest) [inline src/OpenTelemetry.Contrib.Instrumentation.MySqlData/MySqlDataInstrumentation.cs:77]: I'm not familiar with the functionality of the `MySqlTrace` class, so I don't know what to expect from these different event types. This `NonQuery` event type makes me wonder if inserts/deletes/updates would not be covered by this instrumentation. Do you know if QueryOpened/Closed covers these use cases?

Comment (moonheart) [reply src/OpenTelemetry.Contrib.Instrumentation.MySqlData/MySqlDataInstrumentation.cs:77]: The `NonQuery` event only triggered when `SetDatabase` called, QueryOpened/Closed will cover select/insert/update/delete.

reference:
https://github.com/mysql/mysql-connector-net/blob/e39b64edbb7d2b44ffd05b2ab36a399bed8f24e8/MySQL.Data/src/TracingDriver.cs#L140-L145

**Review by alanwest — APPROVED** [review]
This is pretty awesome @moonheart, thanks for contributing!

I don't think any of the projects in the contrib repo currently have integration tests, but it at some point it might be nice to introduce some. Here's how they've been done using docker-compose  for the SqlClient https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/test/OpenTelemetry.Instrumentation.SqlClient.Tests/docker-compose.yml

**Comment by moonheart** [comment]
@alanwest I made a PR before to the main repo with intergration tests https://github.com/open-telemetry/opentelemetry-dotnet/pull/2045 , but they ask me to submit PR to this repo 😄. I can copythese integration tests to here.

**Comment by cijothomas** [comment]
> @alanwest I made a PR before to the main repo with intergration tests [open-telemetry/opentelemetry-dotnet#2045](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2045) , but they ask me to submit PR to this repo 😄. I can copythese integration tests to here.

You can add integration test as a follow up PR.

**Comment by cijothomas** [comment]
Would you be okay to be listed as an approver for mysql.data here: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CODEOWNERS?

(unfortunately, the approver model is not fully ironed out - even though you are listed as approver, you won't actually have approver right. Its currently more like a "SME" , so you can be mentioned, if there are issues with this instrumentation.

File: .github/workflows/package-Instrumentation.MySqlData.yml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+    - name: Publish Artifacts
+      uses: actions/upload-artifact@v2
+      with:
+        name: ${{env.PROJECT}}-packages
+        path: '**/${{env.PROJECT}}/bin/**/*.*nupkg'
+
+    - name: Publish MyGet
+      run: |
+        nuget setApiKey ${{ secrets.MYGET_TOKEN }} -Source https://www.myget.org/F/opentelemetry-contrib/api/v2/package
```
Comment (cijothomas) [inline .github/workflows/package-Instrumentation.MySqlData.yml:49]: could you model this from the existing files? we recently changed this to publish straight to nuget.org

Comment (moonheart) [reply .github/workflows/package-Instrumentation.MySqlData.yml]: Yes, I've added a new commit for this.

**Comment by moonheart** [comment]
> 
> 
> Would you be okay to be listed as an approver for mysql.data here: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CODEOWNERS?
> 
> (unfortunately, the approver model is not fully ironed out - even though you are listed as approver, you won't actually have approver right. Its currently more like a "SME" , so you can be mentioned, if there are issues with this instrumentation.

Yes, please 😄.

**Comment by cijothomas** [comment]
> > Would you be okay to be listed as an approver for mysql.data here: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CODEOWNERS?
> > (unfortunately, the approver model is not fully ironed out - even though you are listed as approver, you won't actually have approver right. Its currently more like a "SME" , so you can be mentioned, if there are issues with this instrumentation.
> 
> Yes, please 😄.

Please add your name to that list in this PR itself.

**Comment by moonheart** [comment]
> 
> 
> > > Would you be okay to be listed as an approver for mysql.data here: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/CODEOWNERS?
> > > (unfortunately, the approver model is not fully ironed out - even though you are listed as approver, you won't actually have approver right. Its currently more like a "SME" , so you can be mentioned, if there are issues with this instrumentation.
> > 
> > 
> > Yes, please 😄.
> 
> Please add your name to that list in this PR itself.

I've added my name to that file.

**Review by cijothomas — APPROVED** [review]

File: .github/workflows/package-Instrumentation.MySqlData.yml (hunk 6 above)
Comment (BenjaminKeffer) [reply .github/workflows/package-Instrumentation.MySqlData.yml]: When will it be available on nuget.org?

Comment (cijothomas) [reply .github/workflows/package-Instrumentation.MySqlData.yml]: i'll try to push this to myget today.

**Pull request merged**